### PR TITLE
fix: make promise throttle level consistent

### PR DESCRIPTION
**Pull request by peternhale**
@W-13709025@

### CODE REVIEW

File: src/throttledPromiseAll.ts [hunk 1]
```diff
@@ -18,6 +18,15 @@ export type PromiseItem<T, O = T | undefined> = {
   producer: (source: T, throttledPromise: ThrottledPromiseAll<T, O | undefined>) => Promise<O | undefined>;
 };
 
+type IndexedProducer<T, O = T> = PromiseItem<T, O> & {
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:21]: Internal use type to add index to input and output.

Index is used to maintain result order to align results with Promise.all

File: src/throttledPromiseAll.ts [hunk 2]
```diff
@@ -56,7 +65,7 @@ export class ThrottledPromiseAll<T, O = T> {
    * Returns the results of the promises that have been resolved.
    */
   public get results(): Array<O | undefined> {
-    return this.#results;
+    return this.#results.sort((a, b) => (a?.index ?? 0) - (b?.index ?? 0)).map((r) => r?.result);
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:68]: Return the results in same order supplied by user

File: src/throttledPromiseAll.ts [hunk 3]
```diff
@@ -124,14 +133,35 @@ export class ThrottledPromiseAll<T, O = T> {
   }
 
   private async dequeue(): Promise<void> {
-    while (this.queue.length > 0) {
-      const next = this.queue.slice(0, this.concurrency);
-      this.queue.splice(0, this.concurrency);
-      // eslint-disable-next-line no-await-in-loop
-      const results = await Promise.all(
-        next.map((item) => item.producer(item.source, this).catch((e) => Promise.reject(e)))
+    const generator = function* (
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:136]: Makes code a bit more concise

File: src/throttledPromiseAll.ts [hunk 4]
```diff
@@ -124,14 +133,35 @@ export class ThrottledPromiseAll<T, O = T> {
...
+      data: Array<PromiseItem<T, O | undefined>>
+    ): Generator<PromiseItem<T, O | undefined> | undefined> {
+      while (data.length > 0) {
+        yield data.shift();
+      }
+    };
+    const concurrencyPool: Array<Promise<IndexedResult<O> | undefined>> = [];
+    const get = generator(this.queue);
+    let index = 0;
+    while (this.queue.length > 0 || concurrencyPool.length > 0) {
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:146]: Stay here while there are things to do.

File: src/throttledPromiseAll.ts [hunk 5]
```diff
@@ -124,14 +133,35 @@ export class ThrottledPromiseAll<T, O = T> {
...
+      while (concurrencyPool.length < this.concurrency) {
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:147]: Make sure there are always running promises at the concurrency level

File: src/throttledPromiseAll.ts [hunk 6]
```diff
@@ -124,14 +133,35 @@ export class ThrottledPromiseAll<T, O = T> {
...
+        const item = get.next().value as PromiseItem<T, O | undefined>;
+        if (!item) {
+          break;
+        }
+
+        const p: IndexedProducer<T, O> = { ...item, index: index++ };
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:153]: Add in the index property to the supplied producer

File: src/throttledPromiseAll.ts [hunk 7]
```diff
@@ -124,14 +133,35 @@ export class ThrottledPromiseAll<T, O = T> {
...
+        concurrencyPool.push(
+          p
+            .producer(item.source, this)
+            .then((result) => ({ index: p.index, result }))
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:157]: capture the index in the result

File: src/throttledPromiseAll.ts [hunk 8]
```diff
@@ -124,14 +133,35 @@ export class ThrottledPromiseAll<T, O = T> {
...
+            .catch((e) => Promise.reject(e))
+        );
+      }
+      this.#results.push(
+        // eslint-disable-next-line no-await-in-loop
+        await Promise.race([concurrencyPool.shift(), ...concurrencyPool])
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:163]: Race until one finishes.

File: src/throttledPromiseAll.ts [hunk 9]
```diff
@@ -124,14 +134,43 @@ export class ThrottledPromiseAll<T, O = T> {
...
-      const next = this.queue.slice(0, this.concurrency);
-      this.queue.splice(0, this.concurrency);
+    const generator = function* (
+      data: Array<PromiseItem<T, O | undefined>>
+    ): Generator<PromiseItem<T, O | undefined> | undefined> {
+      while (data.length > 0) {
+        yield data.shift();
+      }
+    };
+    const concurrencyPool: Map<number, Promise<IndexedResult<O> | undefined>> = new Map<
```
Comment (peternhale) [inline src/throttledPromiseAll.ts:144]: Make pool a map, which simplifies the removal of pool entries that have resolved.

**Comment by mshanemc** [comment]
QA notes:

checked using plugin-dev's `audit messages` cmd
✅ finds extra messages
✅ finds references to non-existent messages

**Review by mshanemc — APPROVED** [review]

File: src/throttledPromiseAll.ts [hunk 10]
```diff
@@ -56,7 +65,8 @@ export class ThrottledPromiseAll<T, O = T> {
    * Returns the results of the promises that have been resolved.
    */
   public get results(): Array<O | undefined> {
-    return this.#results;
+    const results = this.#results;
```
[the remark this answers is not in the document]
Comment (peternhale) [reply src/throttledPromiseAll.ts:68]: I'll fix it. Debug junk left over

**Pull request merged**